File: src/infrastructure/tools/file/read-file.ts
/**
 * read_file — 读取文件内容（支持行范围、多文件批量、超长行截断）
 *
 * 底层调用 Rust `read_file_with_hash`，返回 content + hash10（供 edit_file 做冲突检测）。
 */
import { invoke } from '@tauri-apps/api/core'
import { withCancelResult } from '@/utils/withCancel'
import { t } from '@/ui/i18n'
import { toolRegistry } from '@/domain/tools'
import type { ToolContext, ToolExecutor, ToolResult } from '@/domain/tools/types'
import { securityService } from '@/services/security-service'
import { formatSize } from './common'
// 截断必须代理对安全，否则 emoji 会被切成孤立代理（IPC 落库/桥接直接报错）
import { sliceHead } from '@/utils/text'

/** Rust read_file 返回类型 */
interface FileReadResult {
  content: string
  hash10: string
  line_count: number
  byte_size: number
}

toolRegistry.register(
    'read_file',
    (async (args: Record<string, any>, ctx: ToolContext): Promise<ToolResult> => {
    const maxLines = +(args.max_lines as number) || 2000
    const maxLineChars = +(args.max_line_chars as number) || 2000
    const startLine = Math.max(0, +(args.start_line as number) || 1)

    // 读取单个文件的核心逻辑（供 paths 批量复用）
    async function readSingleFile(
      rawPath: string,
    ): Promise<{ content: string; uiData: any }> {
      const fullPath = await securityService.resolveSafePath(
        rawPath,
        'r',
        ctx.sessionId,
      )
      const result: FileReadResult = await withCancelResult(
        ctx.abortSignal,
        invoke('read_file_with_hash', { path: fullPath }),
        () =>
          ({
            content: '',
            hash10: '',
            line_count: 0,
            byte_size: 0,
          }) as FileReadResult,
      )

      if (!result.hash10) {
        throw '[Cancelled] File read was cancelled.'
      }

      const lines = result.content.split('\n')
      const totalLines = lines.length
      const startIdx = Math.max(0, startLine - 1)

      const slice: string[] = []
      let truncatedLineCount = 0

      for (let i = startIdx; i < totalLines && slice.length < maxLines; i++) {
        const line = lines[i]
        if (line.length > maxLineChars) {
          const omitted = line.length - maxLineChars
          slice.push(
            `${sliceHead(line, maxLineChars)} … [truncated, ${omitted} chars omitted]`,
          )
          truncatedLineCount++
        } else {
          slice.push(line)
        }
      }

      const displayStart = startIdx + 1
      const displayEnd = startIdx + slice.length
      const remainingLines = Math.max(0, totalLines - displayEnd)

      // ⚠️ 模型侧固定英文（P4b/D2-A）；UI 侧走 uiData 由组件渲染。
      // 文案与 Rust `native_tools/file/read_file.rs` 一致（铁律 1）。
      const headerLines = [
        `📄 ${fullPath}`,
        `📝 ${totalLines} lines / ${formatSize(result.byte_size)}`,
        `🔑 hash10: ${result.hash10}`,
        startLine > 1
          ? `🔢 Showing: lines ${displayStart}-${displayEnd} (total ${totalLines} lines)`
          : `🔢 Showing: lines 1-${displayEnd} (total ${totalLines} lines)`,
      ]

      if (truncatedLineCount > 0) {
        headerLines.push(
          `💡 Tip: ${truncatedLineCount} line(s) too long, truncated to ${maxLineChars} chars per line. Increase max_line_chars to read more`,
        )
      }
      if (startIdx > 0) {
        headerLines.push(`💡 Tip: use start_line=${displayEnd + 1} to read more`)
      }
      if (remainingLines > 0) {
        headerLines.push(
          `💡 Tip: content truncated, ${remainingLines} lines remaining. Use start_line=${displayEnd + 1} to read more`,
        )
      }

      const displayedContent = slice.join('\n')
      return {
        content: headerLines.join('\n') + '\n\n' + displayedContent,
        uiData: {
          content: displayedContent,
          hash10: result.hash10,
          line_count: result.line_count,
          byte_size: result.byte_size,
          fullPath,
          startLine: displayStart,
          endLine: displayEnd,
        },
      }
    }

    try {
      // 支持 paths 数组（批量读取多个文件）
      const paths: string[] = Array.isArray(args.paths)
        ? (args.paths as any[]).filter(
            (p): p is string => typeof p === 'string' && p.trim() !== '',
          )
        : []

      if (paths.length > 0) {
        const results: { content: string; uiData: any }[] = []
        const errors: string[] = []

        for (const p of paths) {
          try {
            const r = await readSingleFile(p)
            results.push(r)
          } catch (e: any) {
            errors.push(`${p} — ${e.message || String(e)}`)
          }
        }

        if (results.length === 0 && errors.length > 0) {
          throw errors.join('\n')
        }

        // 文件之间用分隔线隔开
        const parts = results.map((r, i) =>
          i === 0 ? r.content : '\n---\n' + r.content,
        )
        if (errors.length > 0) {
          parts.push(
            `\n\n⚠️ Failed to read ${errors.length} file(s):\n` +
              errors.map((e) => `  - ${e}`).join('\n'),
          )
        }

        const uiData =
          results.length === 1
            ? results[0].uiData // 单文件 → 保持原有 uiData 结构
            : { files: results.map((r) => r.uiData) }

        return {
          content: parts.join('\n'),
          uiData,
        }
      }

      // 单文件路径（向后兼容）
      const path = args.path as string
      if (!path) {
        throw 'Missing required parameter: "path" or "paths"'
      }
      const r = await readSingleFile(path)
      return { content: r.content, uiData: r.uiData }
    } catch (e: any) {
      throw `Error: failed to read file — ${e.message || String(e)}`
    }
  }) as ToolExecutor,
    t('读取文件'),
)

File: src-tauri/src/agent/native_tools/file/read_file.rs
//! `read_file` 工具（原生）— 读取单个文件或 `paths` 批量读取，带行范围与 uiData。

use crate::agent::native_tools::common::{arg_i64, arg_str, arg_str_array, resolve_safe_path};
use crate::agent::native_tools::{NativeToolCtx, NativeToolOutcome};
use crate::file_ops;
use serde_json::{json, Value};

use super::common::format_size;

/// 读取单个文件并返回格式化内容 + uiData（供 read_file_tool 复用）
async fn read_single_file(
    ctx: &NativeToolCtx<'_>,
    path: &str,
    max_lines: usize,
    start_line: usize,
) -> Result<(String, Value), String> {
    let full_path = resolve_safe_path(path, "r", ctx.security)?;
    let full_path_c = full_path.clone();
    let result = tokio::task::spawn_blocking(move || file_ops::read_file(&full_path_c))
        .await
        .map_err(|e| format!("Task join error: {}", e))?
        .map_err(|e| format!("Error: failed to read file — {}", e))?;

    let lines: Vec<&str> = result.content.split('\n').collect();
    let total_lines = lines.len();
    let start_idx = (start_line - 1).min(total_lines);
    let end_idx = (start_idx + max_lines).min(total_lines);
    let slice = &lines[start_idx..end_idx];

    let display_start = start_idx + 1;
    let display_end = end_idx;

    let mut header = vec![
        format!("📄 {}", full_path),
        format!("📝 {} lines / {}", total_lines, format_size(result.byte_size)),
        format!("🔑 hash10: {}", result.hash10),
        format!("🔢 Showing: lines {}-{} (total {} lines)", display_start, display_end, total_lines),
    ];
    if start_idx > 0 {
        header.push(format!("💡 Tip: use start_line={} to read more", display_end + 1));
    }
    if display_end < total_lines {
        header.push(format!(
            "💡 Tip: content truncated, {} lines remaining. Use start_line={} to read more",
            total_lines - display_end,
            display_end + 1
        ));
    }

    let displayed = slice.join("\n");
    let content = format!("{}\n\n{}", header.join("\n"), displayed);
    let ui_data = json!({
        "content": displayed,
        "hash10": result.hash10,
        "line_count": result.line_count,
        "byte_size": result.byte_size,
        "fullPath": full_path,
        "startLine": display_start,
        "endLine": display_end,
    });
    Ok((content, ui_data))
}

pub(crate) async fn read_file_tool(
    ctx: &NativeToolCtx<'_>,
    args: &Value,
) -> Result<NativeToolOutcome, String> {
    // 与 JS `+(max_lines) || 2000` 一致：缺失或 <=0 时取 2000
    let max_lines = match arg_i64(args, "max_lines") {
        Some(n) if n > 0 => n as usize,
        _ => 2000,
    };
    let start_line = arg_i64(args, "start_line").unwrap_or(1).max(1) as usize;

    // 支持 paths 数组（批量读取多个文件）
    let paths = arg_str_array(args, "paths");
    if !paths.is_empty() {
        let mut contents: Vec<String> = Vec::new();
        let mut ui_files: Vec<Value> = Vec::new();
        let mut errors: Vec<String> = Vec::new();

        for p in &paths {
            match read_single_file(ctx, p, max_lines, start_line).await {
                Ok((content, ui_data)) => {
                    contents.push(content);
                    ui_files.push(ui_data);
                }
                Err(e) => {
                    errors.push(format!("{} — {}", p, e));
                }
            }
        }

        // 组装结果
        let mut parts: Vec<String> = Vec::new();
        if contents.is_empty() && !errors.is_empty() {
            // 全部失败
            return Ok(NativeToolOutcome::error(errors.join("\n")));
        }
        // 文件之间用分隔线隔开
        for (i, c) in contents.iter().enumerate() {
            if i > 0 {
                parts.push("\n---".to_string());
            }
            parts.push(c.clone());
        }
        if !errors.is_empty() {
            parts.push(format!(
                "\n\n⚠️ Failed to read {} file(s):\n{}",
                errors.len(),
                errors.iter().map(|e| format!("  - {}", e)).collect::<Vec<_>>().join("\n")
            ));
        }

        let ui_data = if ui_files.len() == 1 {
            // 单文件 → 保持原有 uiData 结构（向后兼容）
            ui_files.into_iter().next().unwrap()
        } else {
            // 多文件 → uiData.files 数组
            json!({ "files": ui_files })
        };

        return Ok(NativeToolOutcome::Value {
            content: parts.join("\n"),
            ui_data: Some(ui_data),
        });
    }

    // 单文件路径（向后兼容）
    let path = arg_str(args, "path").unwrap_or_default();
    if path.is_empty() {
        return Err("Missing required parameter: \"path\" or \"paths\"".to_string());
    }
    let (content, ui_data) = read_single_file(ctx, &path, max_lines, start_line).await?;
    Ok(NativeToolOutcome::Value {
        content,
        ui_data: Some(ui_data),
    })
}
